File: src/main/java/org/warp/picalculator/math/rules/ExponentRule17.java
package org.warp.picalculator.math.rules;

import org.warp.picalculator.Error;
import org.warp.picalculator.math.Function;
import org.warp.picalculator.math.MathContext;
import org.warp.picalculator.math.functions.Expression;
import org.warp.picalculator.math.functions.Multiplication;
import org.warp.picalculator.math.functions.Number;
import org.warp.picalculator.math.functions.Power;
import org.warp.picalculator.math.functions.Root;

import it.unimi.dsi.fastutil.objects.ObjectArrayList;

/**
 * Exponent rule<br>
 * <b>a√x=x^1/a</b>
 * 
 * @author Andrea Cavalli
 *
 */
public class ExponentRule17 {

	public static boolean compare(Function f) {
		if (f instanceof Root) {
			final Root fnc = (Root) f;
			if (fnc.getParameter1().equals(fnc.getParameter2())) {
				return true;
			}
		}
		return false;
	}

	public static ObjectArrayList<Function> execute(Function f) throws Error {
		final MathContext root = f.getMathContext();
		final ObjectArrayList<Function> result = new ObjectArrayList<>();
		final Multiplication fnc = (Multiplication) f;
		final Function a = fnc.getParameter1();
		final Expression expr = new Expression(root, a);
		final Number two = new Number(root, 2);
		final Power p = new Power(fnc.getMathContext(), expr, two);
		result.add(p);
		return result;
	}

}

File: src/main/resources/rules/ExponentRule17.js
//Imports


var Error = org.warp.picalculator.Error;
var Function = org.warp.picalculator.math.Function;
var MathContext = org.warp.picalculator.math.MathContext;
var Multiplication = org.warp.picalculator.math.functions.Multiplication;
var Number = org.warp.picalculator.math.functions.Number;
var Power = org.warp.picalculator.math.functions.Power;
var Root = org.warp.picalculator.math.functions.Root;

var ObjectArrayList = Java.type("it.unimi.dsi.fastutil.objects.ObjectArrayList");

/**
 * Exponent rule
 * a√x=x^1/a
 * 
 * @author Andrea Cavalli
 *
 */
var rule = {
	// Rule name
	getRuleName: function() {
		return "ExponentRule17";
	},
	// Rule type
	getRuleType: function() {
		return RuleType.EXPANSION;
	},
	/* Rule function
	   Returns:
	     - null if it's not executable on the function "f"
	     - An ObjectArrayList<Function> if it did something
	*/

	execute: function(f) {
		var isExecutable;
		if (ScriptUtils.instanceOf(f, Root.class)) {
			var fnc = f;
			if (fnc.getParameter1().equals(fnc.getParameter2())) {
				isExecutable = true;
			}
		}
	
		if (isExecutable) {
			var root = f.getMathContext();
			var result = new ObjectArrayList();
			var fnc = f;
			var a = fnc.getParameter1();
			var two = new Number(root, 2);
			var p = new Power(fnc.getMathContext(), a, two);
			result.add(p);
			return result;
		} else {
			return null;
		}
	}
}

//Add this rule to the list of rules
RulesManager.addRule(engine.getInterface(rule, Rule.class));